pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,450 @@
+	case stoppedStatus:
+		startRequest := ecs.CreateStartInstanceRequest()
+		startRequest.InstanceId = foundInstance.InstanceId
+
+		_, err = client.StartInstance(startRequest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to start instance %v: %v", foundInstance.InstanceId, err)
+		}
+		return nil, fmt.Errorf("instance %v is in a stopped state", foundInstance.InstanceId)
+	case runningStatus:
+		if len(foundInstance.PublicIpAddress.IpAddress) == 0 {
+			ipAddress := ecs.CreateAllocatePublicIpAddressRequest()
+			ipAddress.InstanceId = foundInstance.InstanceId
+
+			_, err = client.AllocatePublicIpAddress(ipAddress)
+			if err != nil {
+				return nil, fmt.Errorf("failed to allocate ip address for instance %v: %v", foundInstance.InstanceId, err)
+			}
+
+		}
+		return &alibabaInstance{instance: foundInstance}, nil
+	}
+
+	return nil, fmt.Errorf("instance %v is not ready", foundInstance.InstanceId)
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (config string, name string, err error) {
+	return "", "", nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client, err := getClient(c.RegionID, c.AccessKeyID, c.AccessKeySecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get alibaba client: %v", err)
+	}
+
+	createInstanceRequest := ecs.CreateCreateInstanceRequest()
+	createInstanceRequest.ImageId, _ = p.getImageIDForOS(machine.Spec, pc.OperatingSystem)
+	createInstanceRequest.InstanceName = machine.Name
+	createInstanceRequest.InstanceType = c.InstanceType
+	createInstanceRequest.VSwitchId = c.VSwitchID
+	createInstanceRequest.InternetMaxBandwidthOut = requests.Integer(c.InternetMaxBandwidthOut)
+	encodedUserData := base64.StdEncoding.EncodeToString([]byte(userdata))
+	createInstanceRequest.UserData = encodedUserData
+	createInstanceRequest.SystemDiskCategory = "cloud_efficiency"
+	createInstanceRequest.ZoneId = c.ZoneID
+
+	_, err = client.CreateInstance(createInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create instance at Alibaba cloud: %v", err)
+	}
+
+	if err = data.Update(machine, func(updatedMachine *v1alpha1.Machine) {
+		if !kuberneteshelper.HasFinalizer(updatedMachine, finalizerInstance) {
+			updatedMachine.Finalizers = append(updatedMachine.Finalizers, finalizerInstance)
+		}
+	}); err != nil {
+		return nil, fmt.Errorf("failed updating machine %v finzaliers: %v", machine.Name, err)
+	}
+
+	foundInstance, err := getInstance(client, machine.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get alibaba instance %v due to %v", machine.Name, err)
+	}
+
+	tag := ecs.AddTagsTag{
```

Comment:
Please configure the tags as part of the creation by setting it on the `createInstanceRequest` to make this atomic. Otherwise if the creation succeeds at the labeling does not, we won't find our instance in the `Get` func and will orphan it